idl.md

```diff
@@ -0,0 +1,58 @@
+
+# WebRTC UWP and IDL
+
+## Overview
+
+For the WebRTC UWP project, we've decided to use [IDL](https://en.wikipedia.org/wiki/IDL_specification_language) (Interface Definition Language) to auto-generate the wrapper code we need to expose the internal C++ APIs in other language bindings we wish to expose to developers externally. IDL defines a set of objects, structures, lists, maps, sets, dictionaries, properties, exceptions, events, and asynchronous promises in a "header" only style simple language.
+
+## Rationale
+
+Specifically for the WebRTC project, our target was C#, JS, and .NET. To do this on Universal Windows requires wrapping each API in C++ code. Originally our APIs were manually written in C++/CX (a Microsoft only C++ extension language), but maintaining the code base against a moving API set (i.e. WebRTC internal code) while not messing up the rules to convert to and from binded objects proved challenging.
+
+The manual bindings were often subject to copy and paste errors and required heavy lifting every time a new API would be introduced or changed.
+
+## Move to IDL for C++/WinRT
+
+With the introduction of [C++/WinRT](https://github.com/Microsoft/cppwinrt), all the old C++/CX code would need to be discarded and replaced with a new implementation. Unlike C++/CX, C++/WinRT requires only a modern, standards-compliant C++17 compiler and uses no special language extensions. The C++ tooling provided by Microsoft takes [Microsoft IDL](https://docs.microsoft.com/en-us/windows/desktop/midl/midl-start-page) and generates a binary `.winmd` file (Windows metadata) describing the entire API surface available. Another tool takes the `.winmd` file and generates C++ stubs matching the API description which are then filled out with an implementation.
+
+### Why not use Microsoft IDL?
+
+The WebRTC UWP project does not use Microsoft IDL directly, but instead uses "SW" IDL input files. These IDL files are consumed by a tool which generates Microsoft IDL as well as all the implementation needed for the C++/WinRT stub files and spits out glue code that needs to be implemented between the exposed API language (in this case C++/WinRT) and the native API implementation.

```

Comment:
Not sure about the "sw" branding. For now, it's "zs".

Reply:
Done